### DISCUSSION TITLE
Support giving a `minimum` & `maximum` to `Data.Floats`

### DISCUSSION THREAD

**Discussion by Seelengrab** [Seelengrab/Supposition.jl#22] [Design & Ideas]
This is currently not supported since the data model is not rich enough to provide good shrinks (without rejecting too many test cases), but would be very nice to have. This would mostly copy what Hypothesis is doing here, which might require #25 so that the generated values continue to shrink well.

The workaround for now is to use `assume!` in user code involving `Data.Floats` to reject any unwanted samples.

**Comment by raineszm** [comment]
This is probably the biggest feature missing from Supposition for me. 

Are there any code contributions that would make it easier to move towards this?

Thanks!

**Reply by Seelengrab** [reply]
Yeah, I figured for numerics people this is going to be much more important  :sweat_smile:

> Are there any code contributions that would make it easier to move towards this?

So #25 is one way that this would be much easier to implement, since there's basically two steps involved:

 1) Generate a `Float64` that shrinks well
 2) Implement clamping of the generated numbers

If you don't want to wait for that, you can take a look at [`lex_to_float`](https://github.com/HypothesisWorks/hypothesis/blob/aad70fb2d9dec2cef9719cdf5369eec9fae0d2a4/hypothesis-python/src/hypothesis/internal/conjecture/floats.py#L176) in Hypothesis, which is the part handling 1). The general idea that Hypothesis is using is drawing a random `Int64` and mapping that in a special way to all bitpatterns of `Float64` so that they shrink well (something `Data.Floats` currently doesn't do well either - this currently just `reinterpret`s, which is far from ideal). 

For 2), the drawn number is clamped in [`draw_float`](https://github.com/HypothesisWorks/hypothesis/blob/aad70fb2d9dec2cef9719cdf5369eec9fae0d2a4/hypothesis-python/src/hypothesis/internal/conjecture/data.py#L1529-L1544) using a function defined [here](https://github.com/HypothesisWorks/hypothesis/blob/aad70fb2d9dec2cef9719cdf5369eec9fae0d2a4/hypothesis-python/src/hypothesis/internal/floats.py#L119). It's this clamper that's the most relevant part here, which would be added directly to [`produce!`](https://github.com/Seelengrab/Supposition.jl/blob/3b47f8b4e59cf8a2ae86f957be633860e35bef04/src/data.jl#L1369), with the bounds added to the struct.

This code in Hypothesis is pretty old and likely doesn't match the way it'd be done in Supposition.jl (e.g. instead of `assert` & throwing an error we `reject(tc)`). The two parts are likely going to be mostly independent, so if you want to give it a try, I'd be happy to help :)

**Reply by Seelengrab** [reply]
Alright, the clamping is implemented on `main` now. It was a bit trickier than I originally thought once I looked it over, so I went ahead and took care of the edge cases. It should be part of the next release.

**Reply by raineszm** [reply]
Wow. That was fast!

**Reply by Seelengrab** [reply]
Thank you! Mind you, this is just 2). If you like, you can still give 1) a try. You'd add that [here](https://github.com/Seelengrab/Supposition.jl/blob/1bb7363586e3348f1467b2928fbe58267fa8a354/src/data.jl#L1454), instead of the `reinterpret`. There's a bunch of floating point utilities in [util.jl](https://github.com/Seelengrab/Supposition.jl/blob/main/src/util.jl) that should make it easy to pull the number apart & put it back together. As for a test that shows the difference - I haven't thought of one yet! I think it'd basically converge faster, so maybe something could be tested there? That'd probably require keeping the current implementation around in the testsuite.

**Reply by raineszm** [reply]
I was looking at doing just that. Perhaps it would a test could check that the shrinking works as expected, i.e. shrinking the input integer shrinks the float in the corresponding way.

Do you have a preference for where the supporting functions like `decode_exponent` should go? In `utils.jl` or `data.jl` or a new file?

**Reply by Seelengrab** [reply]
> i.e. shrinking the input integer shrinks the float in the corresponding way.

That's not directly exposed/accessible - you could check how `SuppositionReport.final_state.calls` compares when running the two side-by-side with the same RNG seed. Whether this will actually improve or not is I guess something that we'll find out afterwards :sweat_smile: 

One complication might be that there's already a bit of floating point special handling just for shrinking in shrink.jl  - see `shrink_float`. We'll have to see how that interacts.

> Do you have a preference for where the supporting functions like decode_exponent should go? In utils.jl or data.jl or a new file?

You can chuck the lookup table right above `produce!` for `Floats`. I don't think you'll need the dedicated function, just indexing should be fine. Note that you likely won't need `update_mantissa`; `assemble` from `utils.jl` should already be able to handle the various manipulations you need to do. IIRC, the IR they've chosen matches the layout of a floating point value exactly, so that should work out fine.  It's easier to work on the parts of the floating point directly instead of manually hacking around with shifts and such, and that also allows extending this to `Float32` and `Float16` in one fell swoop.

For other utility functions (e.g. things like `max_exponent`) that would need to be extended to more than `Float64`, putting them in util.jl is good :+1: I'll probably reorganize all of data.jl at some point anyway, it's starting to get quite unwieldy.

**Reply by raineszm** [reply]
Cool. I'm working on implementing that. We'll see if I get it figured out in a reasonable amount of time; this is my first time mucking about with floating point encodings.

What I meant for tests was that the encoding used by hypothesis is supposed to mean that smaller integers -> 'simpler' floats which is testable. I know the encoding table handles this for the exponents but I thought update_mantissa did a bit of fiddling with the encoding to make this true also for fractional bits.

**Reply by Seelengrab** [reply]
> What I meant for tests was that the encoding used by hypothesis is supposed to mean that smaller integers -> 'simpler' floats which is testable.

Right, what I mean is that Hypothesis does that by having access to an object dedicated to shrinking int -> float; Supposition.jl has an analog in `shrink_float` (see `src/shrink.jl:93`), but that requires having access to both a `TestState` and an `Attempt`. That makes this kind of test much more complicated than it needs to be, so I think we can postpone that until I get to finish the implementation of #25.

There's a bunch of other good tests in the hypothesis tests though that don't require this kind of access, so porting those would be good!

> I know the encoding table handles this for the exponents but I thought update_mantissa did a bit of fiddling with the encoding to make this true also for fractional bits.

Most of `update_mantissa` is fiddling around with just the extraction of the fractional part, which is the part that the functions in `util.jl` take care of for you generically ;) The `reverse_bits` function should be implementable with `bitreverse` and/or `bitrotate` from Base.
